src/Agent.Worker/Build/FileContainerServer.cs

```diff
@@ -337,12 +311,135 @@ private async Task<UploadResult> UploadAsync(IAsyncCommandContext context, int u
             return new UploadResult(failedFiles, uploadedSize);
         }
 
-        private async Task<(DedupIdentifier dedupId, ulong length)> UploadToBlobStore(IAsyncCommandContext context, string itemPath, CancellationToken cancellationToken)
+        private async Task<UploadResult> BlobUploadAsync(IAsyncCommandContext context, IReadOnlyList<string> files, int concurrentUploads, CancellationToken token)
         {
+            // return files that fail to upload and total artifact size
+            var uploadResult = new UploadResult();
+
+            // nothing needs to upload
+            if (files.Count == 0)
+            {
+                return uploadResult;
+            }
+
             var verbose = String.Equals(context.GetVariableValueOrDefault("system.debug"), "true", StringComparison.InvariantCultureIgnoreCase);
+            var (dedupClient, clientTelemetry) = await DedupManifestArtifactClientFactory.Instance
+                .CreateDedupClientAsync(verbose, (str) => context.Output(str), this._connection, token);
+
+            // Upload to blobstore
+            var results = await BlobStoreUtils.UploadBatchToBlobstore(verbose, files, (level, uri, type) =>
```

Comment:
Looks like we are not adding the reference to the top most root. I think it should be fine since we are adding individual root refs to the files, not entirely sure on whether DedupGC would clean up all it's children.

Reply:
I can't imagine the GC would clean up nodes with references. Brian seemed to confirm this.